as.c

```diff
@@ -32,11 +33,338 @@
+        swprintf(out_name, out_name_length, L"%ls.ovpn", profile_name);
+    else
+        swprintf(out_name, out_name_length, L"%ls.ovpn", url);
+
+    out_name[out_name_length - 1] = L'\0';
+
+    /* sanitize profile name */
+    while (*out_name) {
+        wchar_t c = *out_name;
+        if (c == L'<' || c == L'>' || c == L':' || c == L'\"' || c == L'/' || c == L'\\' || c == L'|' || c == L'?' || c == L'*')
+            *out_name = L'_';
+        ++out_name;
+    }
+
+    free(buf);
+}
+
+void
+ShowWinInetError(HANDLE hWnd)
+{
+    WCHAR err[256] = { 0 };
+    FormatMessageW(FORMAT_MESSAGE_FROM_HMODULE | FORMAT_MESSAGE_FROM_SYSTEM, GetModuleHandleW(L"wininet.dll"),
+        GetLastError(), MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), err, _countof(err), NULL);
+    ShowLocalizedMsgEx(MB_OK, hWnd, _T(PACKAGE_NAME), IDS_ERR_URL_IMPORT_PROFILE, GetLastError(), err);
+}
+
+struct UrlComponents
+{
+    int port;
+    WCHAR host[URL_LEN];
+    bool https;
+};
+
+/**
+* Extracts protocol, port and hostname from URL
+*
+* @param url URL to parse, length must be less than URL_MAX
+*/
+void
+ParseUrl(const WCHAR *url, struct UrlComponents* comps)
+{
+    ZeroMemory(comps, sizeof(struct UrlComponents));
+
+    int domain_off = 0;
+    if (wcsbegins(url, L"http://")) {
+        comps->port = 80;
+        domain_off = 7;
+    } else if (wcsbegins(url, L"https://")) {
+        comps->port = 443;
+        comps->https = true;
+        domain_off = 8;
+    } else
+        comps->https = true;
+
+    WCHAR* strport = wcsstr(url + domain_off, L":");
+    if (strport) {
+        WCHAR* end;
+        wcsncpy(comps->host, url + domain_off, strport - url - domain_off);
+        comps->port = wcstol(strport + 1, &end, 10);
+    }
+    else {
+        if (comps->port == 0)
+            comps->port = 443;
+        wcscpy(comps->host, url + domain_off);
+    }
+
+    if (comps->host[wcslen(comps->host) - 1] == L'/')
+        comps->host[wcslen(comps->host) - 1] = L'\0';
+}
+
+/**
+ * Download profile content
+ *
+ * @param hWnd handle of window which initiated download
+ * @param hRequest WinInet request handle
+ * @param pbuf pointer to a buffer, will be allocated by this function. Caller must free it after use.
+ * @parap psize pointer to a profile size, assigned by this function
+ */
+BOOL
+DownloadProfileContent(HANDLE hWnd, HINTERNET hRequest, char** pbuf, size_t* psize)
+{
+    size_t pos = 0;
+    size_t size = READ_CHUNK_LEN;
+
+    *pbuf = calloc(1, size + 1);
+    char* buf = *pbuf;
+    if (buf == NULL) {
+        MessageBoxW(hWnd, L"Out of memory", _T(PACKAGE_NAME), MB_OK);
+        return FALSE;
+    }
+    while (true) {
+        DWORD bytesRead = 0;
+        if (!InternetReadFile(hRequest, buf + pos, READ_CHUNK_LEN, &bytesRead)) {
+            ShowWinInetError(hWnd);
+            return FALSE;
+        }
+        if (bytesRead == 0) {
+            size = pos;
+            break;
+        }
+
+        buf[pos + bytesRead] = '\0';
```

Comment:
This should be moved before the break above, isn't it?

Reply:
In my tests `InternetReadFile` added `NULL` terminator to a buffer when `bytesRead` was `0` (and yes I checked that this is the case for `realloc`-allocated memory, not `calloc`). Since this behavior is not mentioned in docs, let's indeed add `NULL` terminator no matter what `bytesRead` value is.